commit b46b83239e46c93bf633732e5ed38745c2a6029a
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Mon Mar 3 15:03:50 2025 -0500
Commit:     GitHub <noreply@github.com>
CommitDate: Tue Mar 4 04:03:50 2025 +0800

    Support data root in workspace and add entity props for dashboard-based provision (#3262)
    
    Fixes # .
    
    ### Description
    
    This PR adds features to support CC:
    
    - Added support of extra data roots in Workspace.
    
    Three more extra roots can be specified with env vars: NVFL_RESULT_ROOT,
    NVFL_LOG_ROOT, NVFL_AUDIT_ROOT.
    
    This allows different types of data to be placed in different file
    systems: startup kit content is under the root_dir (same as now), output
    generated by application is under RESULT_ROOT, log files will be under
    LOG_ROOT, and audit files will be under AUDIT_ROOT. Different file
    systems could be managed with different policies. For example, the
    content in RESULT_ROOT could be encrypted; LOG_ROOT or AUDIT_ROOT could
    be write-only, encrypted or not.
    
    These extra roots are optional. When not specified, everything will be
    under the root_dir (like in current version).
    
    Content under other roots is organized as:  <root>/<site_name>/<job_id>.
    
    - Added methods in Workspace for application to get result root. Changed
    np_trainer to use data root for model output.
    
    - Make the job store root to be configurable using env var
    NVFL_JOB_STORE_ROOT.
    
    - Augmented dashboard SQL schema with extra properties to capture
    additional config parameters. This will be used for storing provision
    properties required by new features (CC, multiple hostnames, connection
    security, etc.).
    
    These extra properties are dict but are stored as strings, which are
    generated by json.dumps(props).
    
    
    ### Types of changes
    <!--- Put an `x` in all the boxes that apply, and remove the not
    applicable items -->
    - [x] Non-breaking change (fix or new feature that would not break
    existing functionality).
    - [ ] Breaking change (fix or new feature that would cause existing
    functionality to change).
    - [ ] New tests added to cover the changes.
    - [ ] Quick tests passed locally by running `./runtest.sh`.
    - [ ] In-line docstrings updated.
    - [ ] Documentation updated.
---
 nvflare/private/fed/client/client_engine.py | 21 ++++++++------------
 nvflare/private/fed/server/job_runner.py    | 30 ++++++++++++++++++++---------
 nvflare/private/fed/server/server_engine.py | 13 ++++---------
 3 files changed, 33 insertions(+), 31 deletions(-)

diff --git a/nvflare/private/fed/client/client_engine.py b/nvflare/private/fed/client/client_engine.py
index 48170ac8..aed568e0 100644
--- a/nvflare/private/fed/client/client_engine.py
+++ b/nvflare/private/fed/client/client_engine.py
@@ -21,14 +21,7 @@ from typing import List
 
 from nvflare.apis.event_type import EventType
 from nvflare.apis.fl_component import FLComponent
-from nvflare.apis.fl_constant import (
-    FLContextKey,
-    MachineStatus,
-    ProcessType,
-    ReservedKey,
-    SystemComponents,
-    WorkspaceConstants,
-)
+from nvflare.apis.fl_constant import FLContextKey, MachineStatus, ProcessType, ReservedKey, SystemComponents
 from nvflare.apis.fl_context import FLContext, FLContextManager
 from nvflare.apis.shareable import Shareable
 from nvflare.apis.streaming import ConsumerFactory, ObjectProducer, StreamableEngine, StreamContext
@@ -326,7 +319,8 @@ class ClientEngine(ClientEngineInternalSpec, StreamableEngine):
     def get_engine_status(self):
         running_jobs = []
         for job_id in self.get_all_job_ids():
-            run_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
+            ws = Workspace(self.args.workspace)
+            run_folder = ws.get_run_dir(job_id)
             app_name = ""
             app_file = os.path.join(run_folder, "fl_app.txt")
             if os.path.exists(app_file):
@@ -357,11 +351,11 @@ class ClientEngine(ClientEngineInternalSpec, StreamableEngine):
         if status == ClientStatus.STARTED:
             return "Client app already started."
 
-        app_root = os.path.join(
+        workspace = Workspace(
             self.args.workspace,
-            WorkspaceConstants.WORKSPACE_PREFIX + str(job_id),
-            WorkspaceConstants.APP_PREFIX + self.client.client_name,
+            site_name=self.client.client_name,
         )
+        app_root = workspace.get_app_dir(job_id)
         if not os.path.exists(app_root):
             return f"{ERROR_MSG_PREFIX}: Client app does not exist. Please deploy it before starting client."
 
@@ -469,7 +463,8 @@ class ClientEngine(ClientEngineInternalSpec, StreamableEngine):
         return ""
 
     def delete_run(self, job_id: str) -> str:
-        job_id_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
+        ws = Workspace(self.args.workspace)
+        job_id_folder = ws.get_run_dir(job_id)
         if os.path.exists(job_id_folder):
             shutil.rmtree(job_id_folder)
         return f"Delete run folder: {job_id_folder}."
diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 3d82271b..46c3a038 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -14,7 +14,6 @@
 
 import os
 import shutil
-import tempfile
 import threading
 import time
 from typing import Dict, List, Tuple
@@ -28,7 +27,6 @@ from nvflare.apis.job_def import ALL_SITES, Job, JobMetaKey, RunStatus
 from nvflare.apis.job_scheduler_spec import DispatchInfo
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.utils.argument_utils import parse_vars
-from nvflare.fuel.utils.zip_utils import zip_directory_to_file
 from nvflare.lighter.utils import verify_folder_signature
 from nvflare.private.admin_defs import Message, MsgHeader, ReturnCode
 from nvflare.private.defs import RequestHeader, TrainingTopic
@@ -382,16 +380,30 @@ class JobRunner(FLComponent):
 
     def _save_workspace(self, fl_ctx: FLContext):
         job_id = fl_ctx.get_prop(FLContextKey.CURRENT_JOB_ID)
-        workspace = Workspace(root_dir=self.workspace_root)
+        workspace = fl_ctx.get_workspace()
         run_dir = workspace.get_run_dir(job_id)
         engine = fl_ctx.get_engine()
         job_manager = engine.get_component(SystemComponents.JOB_MANAGER)
-        with tempfile.TemporaryDirectory() as td:
-            output_file = os.path.join(td, "workspace")
-            zip_directory_to_file(run_dir, "", output_file)
-            job_manager.save_workspace(job_id, output_file, fl_ctx)
-            self.log_debug(fl_ctx, f"Workspace zipped to {output_file}")
-        shutil.rmtree(run_dir)
+        ws_dirs = [run_dir]
+
+        result_root = workspace.get_result_root(job_id)
+        if result_root not in ws_dirs:
+            ws_dirs.append(result_root)
+
+        log_root = workspace.get_log_root(job_id)
+        if log_root not in ws_dirs:
+            ws_dirs.append(log_root)
+
+        audit_root = workspace.get_audit_root(job_id)
+        if audit_root not in ws_dirs:
+            ws_dirs.append(audit_root)
+
+        location = job_manager.save_workspace(job_id, ws_dirs, fl_ctx)
+        self.log_debug(fl_ctx, f"Workspace {ws_dirs} saved to {location}")
+
+        # remove all ws dirs
+        for d in ws_dirs:
+            shutil.rmtree(d)
 
     def run(self, fl_ctx: FLContext):
         """Starts job runner."""
diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index b293b5e9..376a0169 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -126,14 +126,9 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
 
         self.kv_list = parse_vars(args.set)
 
-    def _get_server_app_folder(self):
-        return WorkspaceConstants.APP_PREFIX + SiteType.SERVER
-
-    def _get_client_app_folder(self, client_name):
-        return WorkspaceConstants.APP_PREFIX + client_name
-
     def _get_run_folder(self, job_id):
-        return os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
+        workspace = Workspace(self.args.workspace)
+        return workspace.get_run_dir(job_id)
 
     def get_engine_info(self) -> EngineInfo:
         self.engine_info.app_names = {}
@@ -144,7 +139,7 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
 
         keys = list(self.run_processes.keys())
         for job_id in keys:
-            run_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(job_id))
+            run_folder = self._get_run_folder(job_id)
             app_file = os.path.join(run_folder, "fl_app.txt")
             if os.path.exists(app_file):
                 with open(app_file, "r") as f:
@@ -161,7 +156,7 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
         return None
 
     def delete_job_id(self, num):
-        job_id_folder = os.path.join(self.args.workspace, WorkspaceConstants.WORKSPACE_PREFIX + str(num))
+        job_id_folder = self._get_run_folder(str(num))
         if os.path.exists(job_id_folder):
             shutil.rmtree(job_id_folder)
         return ""
